### PR TITLE
add shifted gaussian feature

### PR DESCRIPTION
**Pull request by yuanzhou0827**
## Description

Add  `ExpandedGaussian` pair potential in `md.pair`.

## Motivation and context
DNA coated particles have interactions at one end while the other end attaches on the particles surface. The Gaussian potential should be shifted based on the length of DNA or other ligands.

## How has this been tested?
I provided the expected potentials and forces and tested it with pytest.

## Change log

```

```

## Checklist:

- [ ] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

### CODE REVIEW

**Comment by tommy-waltmann** [comment]
Is this ready for review? There seem to be some compilation errors. Also for future reference, please fill out the sections in the PR description. It makes it much easier to see and understand the context behind what is being changed.

**Comment by yuanzhou0827** [comment]
Not yet. There are some errors to be fixed. Thanks for the reminder. I will add the description section.

**Comment by joaander** [comment]
Also, the potential should be named `ExpandedGaussian`.

**Comment by yuanzhou0827** [comment]
Okay! I will change the name too.

**Comment by yuanzhou0827** [comment]
pre-commit.ci autofix

**Comment by joaander** [comment]
I merged `trunk-minor` into this branch. It appears ready for review @tommy-waltmann.

File: hoomd/md/pair/pair.py [hunk 1]
```diff
@@ -293,6 +293,63 @@ def __init__(self, nlist, default_r_cut=None, default_r_on=0., mode='none'):
...
+    r"""Expanded Gaussian pair force.
+
+    Args:
+        nlist (hoomd.md.nlist.NeighborList): Neighbor list.
+        default_r_cut (float): Default cutoff radius :math:`[\mathrm{length}]`.
+        default_r_on (float): Default turn-on radius :math:`[\mathrm{length}]`.
+        mode (str): Energy shifting/smoothing mode.
+
+    `ExpandedGaussian` computes the Gaussian pair force should on every particle
+    in the simulation state:
```
Comment (tommy-waltmann) [inline hoomd/md/pair/pair.py]: I'm not sure what this sentence is supposed to be saying. Is the word should supposed to be shifted?

File: hoomd/md/pair/pair.py [hunk 2]
```diff
@@ -293,6 +293,63 @@ def __init__(self, nlist, default_r_cut=None, default_r_on=0., mode='none'):
...
+    Example::
+
+        nl = nlist.Cell()
+        sgauss = pair.ExpandedGaussian(default_r_cut=3.0, nlist=nl)
+        sgauss.params[('A', 'A')] = dict(epsilon=1.0, sigma=1.0, delta=0.5)
+        sgauss.r_cut[('A', 'B')] = 3.0
+
+    .. py:attribute:: params
+
+        The expanded Gauss potential parameters. The dictionary has the
```
Comment (tommy-waltmann) [inline hoomd/md/pair/pair.py]:
```suggestion
        The expanded Gaussian potential parameters. The dictionary has the
```

**Review by tommy-waltmann — CHANGES_REQUESTED** [review]
The changes that are there now look mostly good, but please update the sphinx doc page `sphinx-doc/module-md-pair.rst` so we can see how the documentation for this potential renders.

File: hoomd/md/pair/pair.py (hunk 2 above)
Comment (yuanzhou0827) [reply hoomd/md/pair/pair.py]: I updated the `phinx-doc/module-md-pair.rst` and the related doc part in the `pair/pair.py`.

File: hoomd/md/EvaluatorPairExpandedGaussian.h [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+
+#ifndef __HIPCC__
+#include <string>
+#endif
+
+#include "hoomd/HOOMDMath.h"
+
+/*! \file EvaluatorPairEXPANDEDGAUSSIAN.h
+    \brief Defines the pair evaluator class for shifted Gaussian potentials
+*/
```
Comment (joaander) [inline hoomd/md/EvaluatorPairExpandedGaussian.h]: Not necessary.

File: hoomd/md/EvaluatorPairExpandedGaussian.h [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+    <b>Gauss specifics</b>
+
+    EvaluatorPairExpandedGaussian evaluates the function:
+    \f[ V_{\mathrm{expanded_gauss}}(r) = \varepsilon \exp \left[ -
+    \frac{1}{2}\left( \frac{r-\delta}}{\sigma}\right)^2 \right] \f]
+
+    The expanded Gaussian potential does not need diameter or charge. Three
+    parameters are specified and stored in a Scalar3.
+    \a epsilon is placed in \a params.x, \a sigma is in \a params.y, and
+    \a delta is in \a params.z.
```
Comment (joaander) [inline hoomd/md/EvaluatorPairExpandedGaussian.h]:
```suggestion
    The expanded Gaussian potential does not need diameter or charge.
```
Remove inaccurate comment.

File: hoomd/md/EvaluatorPairExpandedGaussian.h [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+
+    \a epsilon and \a sigma are related to the standard lj parameters sigma and
+    epsilon by:
+    - \a epsilon = \f$ \varepsilon \f$
+    - \a sigma = \f$ \sigma \f$
+
+    \a delta is the shifted distance of the gaussian potential.
+
```
Comment (joaander) [inline hoomd/md/EvaluatorPairExpandedGaussian.h]:
```suggestion
```
Remove comment that applies to EvaluatorPairLJ.

File: hoomd/md/EvaluatorPairExpandedGaussian.h [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+            delta = v["delta"].cast<Scalar>();
+            }
+
+        // used to facilitate unit testing
+        param_type(Scalar eps, Scalar sig, Scalar r, bool managed = false)
+            {
+            sigma = sig;
+            epsilon = eps;
+            delta = r;
+            }
```
Comment (joaander) [inline hoomd/md/EvaluatorPairExpandedGaussian.h]:
```suggestion
```
Not necessary.

File: hoomd/md/EvaluatorPairExpandedGaussian.h [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+            v["delta"] = delta;
+            return v;
+            }
+#endif
+        }
+#ifdef SINGLE_PRECISION
+        __attribute__((aligned(8)));
+#else
+        __attribute__((aligned(16)));
+#endif
```
Comment (joaander) [inline hoomd/md/EvaluatorPairExpandedGaussian.h]:
```suggestion
        __attribute__((aligned(16)));
```
3 floats is 12 bytes - align to 16 byte boundaries.

**Review by joaander — CHANGES_REQUESTED** [review]
Thanks for adding this. I have suggestions to remove unnecessary code and inaccurate comments.

File: sphinx-doc/module-md-pair.rst [hunk 8]
```diff
@@ -54,6 +55,7 @@ md.pair
         Fourier,
         Gauss,
         Gaussian,
+        ExpandedGaussian,
```
Comment (joaander) [inline sphinx-doc/module-md-pair.rst]: Move `ExpandedGaussian` up. These lists are alphabetized.

File: hoomd/md/pair/pair.py [hunk 9]
```diff
@@ -293,6 +293,64 @@ def __init__(self, nlist, default_r_cut=None, default_r_on=0., mode='none'):
...
+
+        The expanded Gaussian potential parameters. The dictionary has the
+        following keys:
+
+        * ``epsilon`` (`float`, **required**) - energy parameter
+          :math:`\varepsilon` :math:`[\mathrm{energy}]`
+        * ``sigma`` (`float`, **required**) - particle size
+          :math:`\sigma` :math:`[\mathrm{length}]`
+        * ``delta`` (`float`, **required**) - shift distabce
+          :math:`\delta` :math:`[\mathrm{length}]`
```
Comment (joaander) [inline hoomd/md/pair/pair.py:330]:
```suggestion
          :math:`\Delta` :math:`[\mathrm{length}]`
```
For consistency with `ExpandedLJ` and `ExpandedMie`.

File: hoomd/md/pair/pair.py [hunk 10]
```diff
@@ -293,6 +293,64 @@ def __init__(self, nlist, default_r_cut=None, default_r_on=0., mode='none'):
...
+        default_r_cut (float): Default cutoff radius :math:`[\mathrm{length}]`.
+        default_r_on (float): Default turn-on radius :math:`[\mathrm{length}]`.
+        mode (str): Energy shifting/smoothing mode.
+
+    `ExpandedGaussian` computes the radially-shifted Gaussian pair force should
+    on every particle in the simulation state:
+
+    .. math::
+        U(r) = \varepsilon \exp \left( -\frac{1}{2}
+               \left( \frac{r-\delta}{\sigma} \right)^2 \right)
```
Comment (joaander) [inline hoomd/md/pair/pair.py]:
```suggestion
               \left( \frac{r-\Delta}{\sigma} \right)^2 \right)
```
For consistency with `ExpandedLJ` and `ExpandedMie`.

File: hoomd/md/EvaluatorPairExpandedGaussian.h [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+            Scalar sigma_sq = sigma * sigma;
+            Scalar rmd_over_sigma_sq = rmd_sq / sigma_sq;
+            Scalar exp_val = fast::exp(-Scalar(1.0) / Scalar(2.0) * rmd_over_sigma_sq);
+
+            force_divr = epsilon / sigma_sq * exp_val * rmd * rinv;
+            pair_eng = epsilon * exp_val;
+
+            if (energy_shift)
+                {
+                pair_eng -= epsilon * fast::exp(-Scalar(1.0) / Scalar(2.0) * rcutsq / sigma_sq);
```
Comment (joaander) [inline hoomd/md/EvaluatorPairExpandedGaussian.h]: This is incorrect. The shift should be a function of (r_cut - delta).

**Comment by joaander** [comment]
@iblanco11981870 we would appreciate your review.

File: hoomd/md/pair/pair.py [hunk 12]
```diff
@@ -293,6 +293,64 @@ def __init__(self, nlist, default_r_cut=None, default_r_on=0., mode='none'):
...
+    .. py:attribute:: params
+
+        The expanded Gaussian potential parameters. The dictionary has the
+        following keys:
+
+        * ``epsilon`` (`float`, **required**) - energy parameter
+          :math:`\varepsilon` :math:`[\mathrm{energy}]`
+        * ``sigma`` (`float`, **required**) - particle size
+          :math:`\sigma` :math:`[\mathrm{length}]`
+        * ``Delta`` (`float`, **required**) - shift distabce
```
Comment (tommy-waltmann) [inline hoomd/md/pair/pair.py]:
```suggestion
        * ``delta`` (`float`, **required**) - shift distance
```

File: hoomd/md/EvaluatorPairExpandedGaussian.h [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+            Scalar rcutmd_over_sigma_sq = rcutmd_sq / sigma_sq;
+
+            if (energy_shift)
+                {
+                pair_eng -= epsilon * fast::exp(-Scalar(1.0) / Scalar(2.0) * rcutmd_over_sigma_sq);
+                }
+            return true;
+            }
+        else
+            return false;
```
Comment (tommy-waltmann) [inline hoomd/md/EvaluatorPairExpandedGaussian.h]:
```suggestion
        else
            {
            return false;
            }
```
I think the hoomd syntax policy is to include open/closed brackets even if the if/else statement is only one line, but I am unsure about this.

**Review by tommy-waltmann — CHANGES_REQUESTED** [review]
Just two small things left to handle. Great work on this @yuanzhou0827!

File: hoomd/md/EvaluatorPairExpandedGaussian.h (hunk 13 above)
Comment (joaander) [reply hoomd/md/EvaluatorPairExpandedGaussian.h]: Yes, this is good practice, though not enforced. We could check and see if `clang-format` has an option to enforce this.

**Review by tommy-waltmann — APPROVED** [review]

**Comment by joaander** [comment]
@iblanco11981870 this has been open 3 weeks, merging without your review. Next time, please let us know if you are not available to review a pull request.

**Review by joaander — APPROVED** [review]

**Pull request merged**